debuginfo-install plugin new tests and improvements

lukash opened the pull request:
A couple general improvements to the behave steps, a slight refactor of the debuginfo-install tests and adding a couple basic tests to that.

I'll be extending the tests for the debuginfo more shortly.

I'm not crazy happy about the <REPOSYNC> marker feature (see the commits), but it does the job I think we need to do somehow. Suggestions for improvement welcome...

I've named the new repository for the tests without the "dnf-ci-" prefix, as I found it redundant, but I can change it.

lukash commented on dnf-behave-tests/features/steps/cmd.py:
```diff
@@ -179,9 +209,35 @@ def then_stdout_is_empty(context):
+    if expected[0] == "<REPOSYNC>":
+        while i < len(found) and (
+                sync_line.fullmatch(found[i].strip())
+                or last_check_line.fullmatch(found[i].strip())):
```
duude... make up your mind!

lukash replied:
If you look it up (I should have mentioned that in the previous comment) W503 and W504 are the opposites of each other and the W504 is the latter one that recommends operators at the start and deprecates W503.

So obviously flake should only check W504, not both :slightly_smiling_face: I haven't figured a way to report that to them.

lukash commented:
I've added the tests for https://github.com/rpm-software-management/dnf-plugins-core/pull/343 to this PR, although it's gotten a bit big...

pkratoch commented on dnf-behave-tests/features/plugins-core/debuginfo-install.feature:
```diff
@@ -0,0 +1,176 @@
+
+
+@bz1585137
+@with_os=rhel__eq__8
```
Should be ``@use.with_os=rhel__eq__8``.

lukash commented:
> Can you add two more tests for the following scenarios, please?
> 
>     1. `dnf debuginfo-install foo-debuginfo` installs only debuginfo and not debugsource (and vice versa, but that's probably not necessary)

But does it? I'll add the test anyway, but currently it installs both. One problem is that by default debugsource gets pulled in as recommended by debuginfo. Another thing is I'm still installing both packages unconditionally in the plugin. If the user wants just the debuginfo like this, they can call `dnf install foo-debuginfo`? (granted they have to enable the debug repo manually as well)

>     2. `dnf debuginfo-install foo` installs debuginfo and debugsource if available even when the base package doesn't exist (it can easily happen with `--repo=*debug*`)

It actually doesn't work (same as before the patch) - unless the package is installed. I guess I should add a test for that... I'm kind of reluctant to do it because it requires a lot of boilerplate code to be added just for this test - a second repository and somehow split the packages between the two...(presumably this would be easier to set up if we had automatic package generation) What do you think? Should I still do it?

> Also, I am thinking about this scenario: I have RPMs `pkg` and `pkg-libs` form SRPM `pkg` and both RPMs have debuginfo. When I run `dnf debuginfo-install pkg-libs`, it will install `pkg-debugsource` and `pkg-libs-debuginfo`. I suppose that's expected and we needn't worry about the case when there is different version of `pkg` and `pkg-libs`, because ideally it shouldn't happen.

I think there is a high probability of some packages actually doing this (probably not for the `-libs` case). But we can't avoid the inconsistency and the version of the argument (or the installed version of the package in arguments) will be what is installed, so it should be fine. Don't have this specific test-case either though...

lukash commented on dnf-behave-tests/features/plugins-core/debuginfo-install.feature:
```diff
@@ -1,12 +1,42 @@
+      """
+  And stderr is
+      """
+      Error: Unable to find a match
```
No? :slightly_smiling_face: The way it is the test passes for me on F30 `dnf-plugins-core`?

pkratoch replied:
Hm, how is that possible? The test fails for me with dnf-plugins-core-4.0.7-1.fc30.noarch and with the upstream version, too. Isn't the problem with the tag? I run the tests on F30 even though there is with_os=rhel__eq__8 (I am not using the dnf-testing.sh script, but rather the run-tests in a container), so isn't it possible that in your case the scenario just gets skipped?

lukash replied:
I am checking the test output, of course...

https://github.com/rpm-software-management/dnf/commit/db5772cd0c3742894f5f2285c5ce6a7dc0f5c0f8

Only took about an hour to find :cry: 

What version of `dnf` do you have? I have `4.2.5.2` which seems to be the latest from updates and still have the bug. Did you install an out-of-repo package perhaps? :slightly_smiling_face: 

How do I find what version of `dnf` is actually on RHEL8?

pkratoch replied:
Oh, I wouldn't have guessed. Good job finding that. I am using dnf-nightly and didn't thought of downgrading dnf to the F30 version, too.

With dnf-4.2.7 it works (prints the spec list). I will find out about RHEL8.

lukash commented on dnf-behave-tests/features/plugins-core/debuginfo-install.feature:
```diff
@@ -2,7 +2,7 @@ Feature: Tests for debuginfo symbols
 @bz1585137
 Scenario: debuginfo-install reports an error when debuginfo is not found
-Given I use the repository "dnf-ci-fedora"
+Given I use the repository "debuginfo-install"
```
Well.. how can it happen that the repositories get mixed up with your system repositories?

Also, the prefix itself (`dnf-ci-`) doesn't say it's a repository. It's just that you know the repositories have this naming convention and nothing else does.

In the context of tests, I find it not useful and redundant. In case we do need to distinguish these from system repositories, then we should probably use it. I just didn't find it necessary.

But I'd rather it was a suffix (so that it's easier to read and work with in directory structure) and maybe it could just be `-ci`?

I've used this as a proposition to change the convention, if we agree on a change we should gradually migrate, otherwise I'll change this to the current convention... We should discuss with @mblaha and others...

lukash commented on dnf-behave-tests/features/steps/cmd.py:
```diff
@@ -150,6 +150,18 @@ def given_enable_plugin(context, plugin):
+@behave.given("I successfully execute rpm with args \"{args}\"")
+def given_i_successfully_execute_rpm_with_args(context, args):
+    when_I_execute_rpm_with_args(context, args)
+    then_the_exit_code_is(context, 0)
```
I guess I'll change it... But I don't like the way Behave works in this regard...

We should just rename all step function names to `step_impl` then, to avoid the illusion the function name matters.

pkratoch commented:
> > ```
> > 1. `dnf debuginfo-install foo-debuginfo` installs only debuginfo and not debugsource (and vice versa, but that's probably not necessary)
> > ```
> 
> But does it? I'll add the test anyway, but currently it installs both. One problem is that by default debugsource gets pulled in as recommended by debuginfo. Another thing is I'm still installing both packages unconditionally in the plugin. If the user wants just the debuginfo like this, they can call `dnf install foo-debuginfo`? (granted they have to enable the debug repo manually as well)

That behavior is fine by me. It's not a regression, because the previous version of ``debuginfo-install`` didn't install debugsource anyway (except for dependencies), so ok.

> > ```
> > 2. `dnf debuginfo-install foo` installs debuginfo and debugsource if available even when the base package doesn't exist (it can easily happen with `--repo=*debug*`)
> > ```
> 
> It actually doesn't work (same as before the patch) - unless the package is installed. I guess I should add a test for that... I'm kind of reluctant to do it because it requires a lot of boilerplate code to be added just for this test - a second repository and somehow split the packages between the two...(presumably this would be easier to set up if we had automatic package generation) What do you think? Should I still do it?

Why would a second repository be needed? Isn't just one ``-debuginfo`` (or two, if you want ``-debugsource``, too) package all we need? Just to be clear, I am not proposing to test the example with  ``--repo=*debug*``, it was just to show how common this problem might be.

lukash commented:
> > It actually doesn't work (same as before the patch) - unless the package is installed. I guess I should add a test for that... I'm kind of reluctant to do it because it requires a lot of boilerplate code to be added just for this test - a second repository and somehow split the packages between the two...(presumably this would be easier to set up if we had automatic package generation) What do you think? Should I still do it?
> 
> Why would a second repository be needed? Isn't just one `-debuginfo` (or two, if you want `-debugsource`, too) package all we need? Just to be clear, I am not proposing to test the example with `--repo=*debug*`, it was just to show how common this problem might be.

You also need the base package, install it and then use a repository with `-debuginfo` that doesn't contain the base package.

As I said, if you don't have the base package at all (seems that is what you are thinking), then the plugin will not work - it needs to search for the base package before it installs the `-debuginfo`. In theory it could just slap the `-debuginfo` suffix to the name, but that falls short quickly, like when you need to find the source package to correctly form the `-debugsource` package...

pkratoch commented:
> As I said, if you don't have the base package at all (seems that is what you are thinking), then the plugin will not work - it needs to search for the base package before it installs the `-debuginfo`. In theory it could just slap the `-debuginfo` suffix to the name, but that falls short quickly, like when you need to find the source package to correctly form the `-debugsource` package...

Ah, I get it now, thanks for explaining and sorry for the confusion. I don't think the test is necessary.

lukash commented:
Ok, pushed a new version with:
- fixed @use.with_os
- added a test for `debuginfo-install foo-debuginfo`
- fixed calling steps
- fixed the `Error: Unable to find a match: non-existent-package` message

So what I think remains is only to discuss the names of the repositories.

lukash commented:
I've pushed a new series which has many more tests for the cases discussed in https://github.com/rpm-software-management/dnf-plugins-core/pull/343. Two of the tests fail atm. (marked with @xfail), still WIP.

I've also added the following three commits:

17d601c1 steps: drop common.string.splitter and use native string split()
36779413 steps: Use print_diff for transactions
87ac2894 Build i686 versions of packages based on their name containing "multilib"

As more general improvements, namely printing a colored diff for transactions and building multilib packages based on their name containing "multilib", instead of having a hard-coded list.

In case the general improvement commits pass review faster, it might be a good idea I create a separate PR for them and we merge them, so that I don't have to maintain a growing list of commits here...

pkratoch commented on dnf-behave-tests/features/steps/common/dnf.py:
```diff
@@ -116,6 +119,9 @@ def parse_transaction_table(lines):
 
     table_begin = find_transaction_table_begin(lines)
+    if table_begin is None:
+        raise AssertionError("Transaction table not found.")
```
Huge amount of tests is failing on this. Mostly probably in case of "Transaction is empty" check, because it checks for ``RuntimeError`` when calling ``parse_transaction_table`` (and it has to try to parse the transaction table, because it's an error if there is some transaction). Can you please fix it?

lukash replied:
Sorry about that, I thought I've run the tests against the changes, but I guess I must have changed this afterwards.

pkratoch commented on dnf-behave-tests/features/steps/transaction.py:
```diff
@@ -58,6 +59,41 @@ def check_dnf_transaction(context, mode):
+        for action, pkgs in dnf_transaction.items():
+            for pkg in pkgs:
+                found_lists.append((action, str(pkg)))
+        found_lists.sort()
```
Unfortunately, we still have the "unchanged" state in some scenarios even when checking for the exact match, so this change requires to adjust these tests, too.

pkratoch replied:
Oh, and similarly, there are "absent" and "present" states. I don't think any of these states bring any value in the "exact_match" mode. Maybe in some cases they are useful for quicker understanding the test scenario, but it can be achieved using comments instead.

lukash replied:
I've actually dropped the transaction diff commit from the PR.

I'll try to drop the whole non-exact mode and all the stuff around it and send another PR. It is just an unnecessary complication while making the tests less reliable at the same time.

lukash commented:
I've pushed a new test for installonly packages.

lukash commented:
Pushed is hopefully the final version of the tests.

pkratoch commented on dnf-behave-tests/features/plugins-core/debuginfo-install.feature:
```diff
@@ -33,10 +35,235 @@ Scenario: reports an error for a package without debuginfo
+      | install       | baz-debugsource-0:2.0-1.x86_64            |
+
+@not.with_os=rhel__eq__8
+Scenario: installs debuginfo for multiple installed architectures
```
Did you mean "versions" instead of "architectures"?

lukash replied:
Copy-paste remnant, thanks.

lukash commented:
I've added more tests for installing `-debuginfo` of the base (parent, source) package when a subpackage has no `-debuginfo` itself.
